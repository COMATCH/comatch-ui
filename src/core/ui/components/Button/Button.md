
## Button

From [`Button.js`](Button.js)



prop | type | default | required | description
---- | :----: | :-------: | :--------: | -----------
**className** | `String` | `''` | :x: | 
**disabled** | `Boolean` | `false` | :x: | 
**ghost** | `Boolean` | `false` | :x: | 
**href** | `String` | `''` | :x: | 
**icon** | `ReactNode` | `null` | :x: | 
**iconAfterText** | `Boolean` | `false` | :x: | 
**id** | `String` | `null` | :x: | 
**onClick** | `Function` | `noop` | :x: | 
**target** | `String` | `null` | :x: | 
**text** | `String` | `''` | :x: | 
**type** | `Enum('button','submit','reset')` | `'button'` | :x: | 



