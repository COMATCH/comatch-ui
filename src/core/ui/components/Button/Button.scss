@import '../../../styles/variables.scss';

.Button {
    border: 1px solid;
    border-radius: 3px;
    display: inline-block;
    font-size: $co-font-sm;
    outline: 0 !important;
    font-size: $co-font-sm;
    touch-action: manipulation;
    cursor: pointer;
    text-align: center;
    text-transform: uppercase;
    white-space: nowrap;
    padding: 8px 15px; // makes the height 40px
    transition: background-color 250ms ease-out, border-color 250ms ease-out, color 250ms ease-out;

    &.full {
        background-color: $co-color-primary;
        border-color: $co-color-primary;
        color: $co-color-white;

        &.disabled {
            background-color: $co-color-light-gray !important;
        }

        &:hover {
            background-color: $co-color-secondary;
            border-color: $co-color-secondary;
        }
    }

    &.ghost {
        background-color: $co-color-white;
        border-color: $co-color-primary;
        color: $co-color-primary;

        &:hover {
            background-color: $co-color-secondary;
            border-color: $co-color-secondary;
            color: $co-color-white;
        }

        &.disabled {
            background-color: $co-color-white !important;
        }
    }

    &.disabled {
        border-color: $co-color-mid-gray !important;
        cursor: not-allowed;
        pointer-events: none;

        &,
        span {
            color: $co-color-mid-gray !important;
        }
    }

    // In case a Button has an icon defined
    svg {
        margin-right: 5px;
    }

    &.only-icon {
        svg {
            margin: 0;
        }
        padding: 8px 12px;
    }

    &.icon-after-text {
        white-space: initial;

        svg {
            float: right;
            position: relative;
            top: 2px;
            margin-left: 5px;
            margin-right: 0;
        }
    }

    // If buttons follow each other in a row
    // set a standardized whitespace between them
    & ~ & {
        margin-left: 9px;
    }
}
