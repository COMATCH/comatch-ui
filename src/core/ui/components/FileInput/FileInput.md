
## FileInput

From [`FileInput.js`](FileInput.js)



prop | type | default | required | description
---- | :----: | :-------: | :--------: | -----------
**accept** | `Union<String \| Array[]<String>>` |  | :white_check_mark: | 
**buttonText** | `String` |  | :white_check_mark: | 
**disabled** | `Boolean` | `false` | :x: | 
**multiple** | `Boolean` | `false` | :x: | 
**name** | `String` | `null` | :x: | 
**onChange** | `Function` |  | :white_check_mark: | 



