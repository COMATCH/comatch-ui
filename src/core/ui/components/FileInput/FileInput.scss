@import '../../../styles/variables.scss';

.FileInput {
    display: inline-block;

    input[type='file'] {
        display: none;
    }
}
