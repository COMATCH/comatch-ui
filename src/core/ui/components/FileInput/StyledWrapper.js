import styled from 'styled-components';

export const StyledWrapper = styled.div`
    display: inline-block;

    input[type='file'] {
        display: none;
    }
`;
