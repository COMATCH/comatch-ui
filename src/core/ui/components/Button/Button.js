import React from 'react';
import PropTypes from 'prop-types';
import noop from 'lodash/noop';
import classNames from 'classnames';

import { StyledWrapperButton, StyledWrapperLink } from './StyledWrapper';
// import './Button.scss';

const propTypes = {
    /**
     * Additional class names
     */
    className: PropTypes.string,
    disabled: PropTypes.bool,
    /**
     * Alternative styling with white background and primary colored text
     */
    ghost: PropTypes.bool,
    /**
     * If true, uses `<a>` instead of `<button>`
     */
    href: PropTypes.string,
    /**
     * HTML id attribute
     */
    id: PropTypes.string,
    /**
     * SVG icon
     */
    icon: PropTypes.node,
    iconAfterText: PropTypes.bool,
    onClick: PropTypes.func,
    /**
     * To be used with `href`, could set for example `target="_blank"`
     */
    target: PropTypes.string,
    /**
     * Text content of `Button`
     */
    text: PropTypes.string,
    type: PropTypes.oneOf(['button', 'submit', 'reset']),
    /**
     * Can have two values: rectangle(by dafauly) and circle. 
     */
    shape: PropTypes.oneOf(['rectangle', 'circle']),
};

const defaultProps = {
    className: '',
    disabled: false,
    ghost: false,
    href: '',
    id: null,
    icon: null,
    iconAfterText: false,
    onClick: noop,
    target: null,
    text: '',
    type: 'button',
    shape: 'rectangle',
};

function renderLinkButton({ classes, content, disabled, href, id, onClick, target, styledProps } = {}) {
    const finalOnClick = disabled ? (evt) => evt.preventDefault() : onClick;
    const rel = target && 'noopener noreferrer';
    return (
        <StyledWrapperLink
            {...styledProps}
            className={classes}
            href={href}
            target={target}
            id={id}
            rel={rel}
            onClick={finalOnClick}>
            {content}
        </StyledWrapperLink>
    );
}

function renderHtmlButton({ classes, content, id, onClick, type, styledProps } = {}) {
    return (
        <StyledWrapperButton
            {...styledProps}
            className={classes}
            id={id}
            type={type}
            onClick={onClick}>
            {content}
        </StyledWrapperButton>
    );
}

/**
 * A `Button` component with different stylistic variations. By default the styling is `full`,
 * but `ghost` is an alternative version with reversed colors.
 * 
 * If `href` is set an `<a>` is used instead of `<button>`. If the `target` prop is set to `_blank`
 * in combination with `href`, then it will also add the prop `rel="noopener noreferrer"` for
 * additional security.
 */

export const Button = ({
    id,
    className,
    href,
    target,
    onClick,

    ghost,
    disabled,

    text,
    type,
    icon,
    iconAfterText,
    shape,
}) => {
    const classes = classNames('Button', className, {
        full: !ghost,
        disabled,
        ghost,
        'only-icon': !text,
        'icon-after-text': text && iconAfterText,
    });
    const content = iconAfterText ? (
        <span>
            {text}
            {icon}
        </span>
    ) : (
        <span>
            {icon}
            {text}
        </span>
    );

    const calculatedProps = {
        id,
        classes,
        href,
        target,

        type,
        content,

        disabled,

        onClick,

        styledProps: {
            disabled,
            ghost,
            full: !ghost,
            onlyIcon: !text,
            iconAfterText: text && iconAfterText,
            shape,
        },
    };

    return href
        ? renderLinkButton(calculatedProps)
        : renderHtmlButton(calculatedProps);
};

Button.propTypes = propTypes;

Button.defaultProps = defaultProps;
