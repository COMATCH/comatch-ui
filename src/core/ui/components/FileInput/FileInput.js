import React, { Component } from 'react';
import PropTypes from 'prop-types';
import set from 'lodash/set';
import { Button } from '../Button';

import './FileInput.scss';

export class FileInput extends Component {
    static propTypes = {
        accept: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)])
            .isRequired,
        buttonText: PropTypes.string.isRequired,
        name: PropTypes.string,
        multiple: PropTypes.bool,
        onChange: PropTypes.func.isRequired,
    };

    static defaultProps = {
        name: null,
        multiple: false,
    };

    constructor(props) {
        super(props);

        this.fileInputRef = React.createRef();
    }

    generateAcceptValue(accept) {
        if (Array.isArray(accept)) {
            return accept.join(',');
        }
        return accept;
    }

    handleButtonClick = () => this.fileInputRef.current.click();

    handleChange = (event) => {
        this.props.onChange(event);
        // After calling the user defined onChange we have to reset the event.target.value
        // to empty string to "trick" the input into triggering another change event in case
        // the user selects the same file they did before for another consecutive time.
        set(event, 'target.value', '');
    }

    render() {
        const { accept, buttonText, multiple } = this.props;

        return (
            <div className="Input FileInput">
                <input
                    ref={this.fileInputRef}
                    type="file"
                    accept={this.generateAcceptValue(accept)}
                    onChange={this.handleChange}
                    multiple={multiple}
                />
                <Button text={buttonText} onClick={this.handleButtonClick} />
            </div>
        );
    }
}
