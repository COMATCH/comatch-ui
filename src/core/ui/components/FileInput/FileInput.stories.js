/* eslint-disable max-len */
import React from 'react';
import Highlight from 'react-highlight';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { FileInput } from './FileInput';

storiesOf('FileInput', module)
    .add('accepting PDFs, single file', () => (
        <>
            <FileInput accept="application/pdf" buttonText="Upload" onChange={action('change')} />
            <Highlight className="html">{`<FileInput accept="application/pdf" buttonText="Upload" onChange={action('change')} />`}</Highlight>
        </>
    ))
    .add('accepting images, single file', () => (
        <>
            <FileInput accept="image/*" buttonText="Upload" onChange={action('change')} />
            <Highlight className="html">{`<FileInput accept="image/*" buttonText="Upload" onChange={action('change')} />`}</Highlight>
        </>
    ))
    .add('accepting images and PDFs, single file', () => (
        <>
            <FileInput accept={['application/pdf', 'image/*']} buttonText="Upload" onChange={action('change')} />
            <Highlight className="html">{`<FileInput accept={['application/pdf', 'image/*']} buttonText="Upload" onChange={action('change')} />`}</Highlight>
        </>
    ))
    .add('accepting images, multiple files', () => (
        <>
            <FileInput accept="image/*" onChange={action('change')} buttonText="Upload" multiple />
            <Highlight className="html">{`<FileInput accept="image/*" onChange={action('change')} buttonText="Upload" multiple />`}</Highlight>
        </>
    ));
