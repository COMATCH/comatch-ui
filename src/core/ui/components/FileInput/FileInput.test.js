import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import '../../../config/tests/setup';

import { FileInput } from './FileInput';
import { Button } from '../Button';

const mockFunction = jest.fn();

const props = {
    accept: '.json',
    buttonText: 'Button Text',
    onChange: mockFunction,
    name: 'Name',
    multiple: true,
};

const acceptArray = ['application/pdf', 'image/*'];

describe('FileInput', () => {
    it('should render correctly', () => {
        const wrapper = renderer.create(<FileInput />).toJSON();

        expect(wrapper).toMatchSnapshot();
    });

    it('should render with default props', () => {
        const wrapper = mount(<FileInput />);

        expect(wrapper.prop('name')).toEqual(null);
        expect(wrapper.prop('multiple')).toEqual(false);
    });

    it('should render with custom props', () => {
        const wrapper = mount(<FileInput {...props} />);

        expect(wrapper.prop('accept')).toEqual('.json');
        expect(wrapper.prop('buttonText')).toEqual('Button Text');
        expect(wrapper.prop('onChange')).toEqual(mockFunction);
        expect(wrapper.prop('name')).toEqual('Name');
        expect(wrapper.prop('multiple')).toEqual(true);
    });

    describe('generateAcceptValue()', () => {
        it('should return a string', () => {
            const wrapper = mount(<FileInput accept={acceptArray} />);
            const input = wrapper.find('input');
            expect(input.prop('accept')).toEqual('application/pdf,image/*');
            expect(typeof input.prop('accept')).toEqual('string');
        });
    });
});
