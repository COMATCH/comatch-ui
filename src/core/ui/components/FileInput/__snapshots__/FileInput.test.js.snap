// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileInput should render correctly 1`] = `
<div
  className="Input FileInput sc-bxivhb jPHirh"
>
  <input
    accept="application/json"
    multiple={false}
    onChange={[Function]}
    type="file"
  />
  <button
    className="Button rectangle full sc-htpNat gpVMmM"
    disabled={false}
    id={null}
    onClick={[Function]}
    type="button"
  >
    <span>
      Button Text
    </span>
  </button>
</div>
`;
